lib/optimizer/src/Error/ErrorProperties.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace Amp\Optimizer\Error;
+
+/**
+ * Default set of properties and methods to use for errors.
+ *
+ * @package Amp\Optimizer
+ */
+trait ErrorProperties
+{
+
+    /**
+     * Instantiate an Error object.
+     *
+     * @param string $message Message for the error.
+     */
+    public function __construct($message)
+    {
+        $this->message = $message;
+    }
+
+    /**
+     * Message of the error.
+     *
+     * @var string
+     */
+    protected $message;
+
+    /**
+     * Get the code of the error.
+     *
+     * @return string Code of the error.
+     */
+    public function getCode()
+    {
+        return basename(get_class($this));
+    }
+
+    /**
+     * Get the message of the error.
+     *
+     * @return string Message of the error.
+     */
+    public function getMessage()
+    {
+        return $this->message;
+    }
```

Comment:
Idea: Should this trait implement `__toString()`? Perhaps:

```php
public function __toString() {
    return $this->getCode() . ': ' . $this->getMessage();
}
```

Would make it a bit easier to log them out.

Reply:
Most of the time, you'd want strict typing, not objects that automatically turn into scalars, as this can happen when you don't want it to.

I'd suggest holding off on this and see how "cumbersome" the usage will end up being.

If the code is properly designed, there wouldn't be that many places where you actually need to output these objects, so we might be talking about one or two occurences of `->getMessage()` only.